// @kissmybutton/motorcortex/dist/DescriptiveIncidents/__tests__/states/dc18.js
"use strict";

module.exports = {
  descriptiveTree: {
    id: 'my-clip',
    duration: 11000,
    children: {
      'my-group': {
        id: 'my-group',
        position: 0,
        duration: 800,
        children: {
          'test-incident-5': {
            id: 'test-incident-5',
            position: 0,
            duration: 800
          }
        }
      },
      'test-incident-2': {
        id: 'test-incident-2',
        position: 5000,
        duration: 6000
      },
      'my-plugin-group': {
        id: 'my-plugin-group',
        position: 4000,
        duration: 300,
        children: {
          'my-plugin-group_0': {
            id: 'my-plugin-group_0',
            duration: 100,
            position: 0
          },
          'my-plugin-group_1': {
            id: 'my-plugin-group_1',
            duration: 100,
            position: 100
          },
          'my-plugin-group_2': {
            id: 'my-plugin-group_2',
            duration: 100,
            position: 200
          }
        }
      },
      'test-group-1': {
        id: 'test-group-1',
        position: 1000,
        duration: 9000,
        children: {
          'test-incident-1': {
            id: 'test-incident-1',
            position: 7000,
            duration: 2000
          }
        }
      }
    }
  },
  lanes: {
    'my-class-1___alpha': [{
      id: 'test-incident-5_my-class-1_alpha',
      millisecond: 0,
      initialValue: 1
    }, {
      id: 'my-plugin-group_0_my-class-1_alpha',
      millisecond: 4000,
      initialValue: 5
    }, {
      id: 'my-plugin-group_1_my-class-1_alpha',
      millisecond: 4100,
      initialValue: 2
    }, {
      id: 'my-plugin-group_2_my-class-1_alpha',
      millisecond: 4200,
      initialValue: 0
    }, {
      id: 'test-incident-1_my-class-1_alpha',
      millisecond: 8000,
      initialValue: 2
    }],
    'my-class-1___beta': [{
      id: 'test-incident-5_my-class-1_beta',
      millisecond: 0,
      initialValue: 0
    }, {
      id: 'test-incident-1_my-class-1_beta',
      millisecond: 8000,
      initialValue: 5
    }],
    'my-class-1___gamma': [{
      id: 'test-incident-2_my-class-1_gamma',
      millisecond: 5000,
      initialValue: 3
    }],
    'my-class-2___alpha': [{
      id: 'test-incident-5_my-class-2_alpha',
      millisecond: 0,
      initialValue: 1
    }, {
      id: 'test-incident-1_my-class-2_alpha',
      millisecond: 8000,
      initialValue: 5
    }],
    'my-class-2___beta': [{
      id: 'test-incident-5_my-class-2_beta',
      millisecond: 0,
      initialValue: 0
    }, {
      id: 'test-incident-1_my-class-2_beta',
      millisecond: 8000,
      initialValue: 5
    }]
  },
  realTree: {
    id: 'my-clip',
    duration: 11000,
    children: {
      'my-group': {
        id: 'my-group',
        position: 0,
        duration: 800,
        children: {
          'test-incident-5': {
            id: 'test-incident-5',
            position: 0,
            duration: 800,
            children: {
              'test-incident-5_my-class-1': {
                id: 'test-incident-5_my-class-1',
                position: 0,
                duration: 800,
                children: {
                  'test-incident-5_my-class-1_alpha': {
                    id: 'test-incident-5_my-class-1_alpha',
                    position: 0,
                    duration: 800
                  },
                  'test-incident-5_my-class-1_beta': {
                    id: 'test-incident-5_my-class-1_beta',
                    position: 0,
                    duration: 800
                  }
                }
              },
              'test-incident-5_my-class-2': {
                id: 'test-incident-5_my-class-2',
                position: 0,
                duration: 800,
                children: {
                  'test-incident-5_my-class-2_alpha': {
                    id: 'test-incident-5_my-class-2_alpha',
                    position: 0,
                    duration: 800
                  },
                  'test-incident-5_my-class-2_beta': {
                    id: 'test-incident-5_my-class-2_beta',
                    position: 0,
                    duration: 800
                  }
                }
              }
            }
          }
        }
      },
      'my-plugin-group': {
        id: 'my-plugin-group',
        position: 4000,
        duration: 300,
        children: {
          'my-plugin-group_0': {
            id: 'my-plugin-group_0',
            position: 0,
            duration: 100,
            children: {
              'my-plugin-group_0_my-class-1': {
                id: 'my-plugin-group_0_my-class-1',
                position: 0,
                duration: 100,
                children: {
                  'my-plugin-group_0_my-class-1_alpha': {
                    id: 'my-plugin-group_0_my-class-1_alpha',
                    position: 0,
                    duration: 100
                  }
                }
              }
            }
          },
          'my-plugin-group_1': {
            id: 'my-plugin-group_1',
            position: 100,
            duration: 100,
            children: {
              'my-plugin-group_1_my-class-1': {
                id: 'my-plugin-group_1_my-class-1',
                position: 0,
                duration: 100,
                children: {
                  'my-plugin-group_1_my-class-1_alpha': {
                    id: 'my-plugin-group_1_my-class-1_alpha',
                    position: 0,
                    duration: 100
                  }
                }
              }
            }
          },
          'my-plugin-group_2': {
            id: 'my-plugin-group_2',
            position: 200,
            duration: 100,
            children: {
              'my-plugin-group_2_my-class-1': {
                id: 'my-plugin-group_2_my-class-1',
                position: 0,
                duration: 100,
                children: {
                  'my-plugin-group_2_my-class-1_alpha': {
                    id: 'my-plugin-group_2_my-class-1_alpha',
                    position: 0,
                    duration: 100
                  }
                }
              }
            }
          }
        }
      },
      'test-incident-2': {
        id: 'test-incident-2',
        position: 5000,
        duration: 6000,
        children: {
          'test-incident-2_my-class-1': {
            id: 'test-incident-2_my-class-1',
            position: 0,
            duration: 6000,
            children: {
              'test-incident-2_my-class-1_gamma': {
                id: 'test-incident-2_my-class-1_gamma',
                position: 0,
                duration: 6000
              }
            }
          }
        }
      },
      'test-group-1': {
        id: 'test-group-1',
        position: 1000,
        duration: 9000,
        children: {
          'test-incident-1': {
            // DOMAwareIncident
            id: 'test-incident-1',
            position: 7000,
            duration: 2000,
            children: {
              'test-incident-1_my-class-1': {
                // AnimatedAttributeSplitter
                id: 'test-incident-1_my-class-1',
                position: 0,
                duration: 2000,
                children: {
                  'test-incident-1_my-class-1_alpha': {
                    // ContextAwareIncidents
                    id: 'test-incident-1_my-class-1_alpha',
                    position: 0,
                    duration: 2000
                  },
                  'test-incident-1_my-class-1_beta': {
                    id: 'test-incident-1_my-class-1_beta',
                    position: 0,
                    duration: 2000
                  }
                }
              },
              'test-incident-1_my-class-2': {
                id: 'test-incident-1_my-class-2',
                position: 0,
                duration: 2000,
                children: {
                  'test-incident-1_my-class-2_alpha': {
                    id: 'test-incident-1_my-class-2_alpha',
                    position: 0,
                    duration: 2000
                  },
                  'test-incident-1_my-class-2_beta': {
                    id: 'test-incident-1_my-class-2_beta',
                    position: 0,
                    duration: 2000
                  }
                }
              }
            }
          }
        }
      }
    }
  }
};
