// @kissmybutton/motorcortex/dist/DescriptiveIncidents/__tests__/states/dc3.js
"use strict";

module.exports = {
  descriptiveTree: {
    id: 'my-clip',
    duration: 7000,
    children: {
      'my-group': {
        id: 'my-group',
        position: 200,
        duration: 0,
        children: {}
      },
      'test-incident-2': {
        id: 'test-incident-2',
        position: 5000,
        duration: 2000
      },
      'test-group-1': {
        id: 'test-group-1',
        position: 1000,
        duration: 3000,
        children: {
          'test-incident-1': {
            id: 'test-incident-1',
            position: 1000,
            duration: 2000
          }
        }
      }
    }
  },
  lanes: {
    'my-class-1___alpha': [{
      id: 'test-incident-1_my-class-1_alpha',
      millisecond: 2000,
      initialValue: 1
    }, {
      id: 'test-incident-2_my-class-1_alpha',
      millisecond: 5000,
      initialValue: 1.1
    }],
    'my-class-1___beta': [{
      id: 'test-incident-1_my-class-1_beta',
      millisecond: 2000,
      initialValue: 0
    }, {
      id: 'test-incident-2_my-class-1_beta',
      millisecond: 5000,
      initialValue: 2.2
    }],
    'my-class-2___alpha': [{
      id: 'test-incident-1_my-class-2_alpha',
      millisecond: 2000,
      initialValue: 1
    }],
    'my-class-2___beta': [{
      id: 'test-incident-1_my-class-2_beta',
      millisecond: 2000,
      initialValue: 0
    }]
  },
  realTree: {
    id: 'my-clip',
    children: {
      'my-group': {
        id: 'my-group',
        position: 200,
        duration: 0,
        children: {}
      },
      'test-incident-2': {
        id: 'test-incident-2',
        position: 5000,
        duration: 2000,
        children: {
          'test-incident-2_my-class-1': {
            id: 'test-incident-2_my-class-1',
            position: 0,
            duration: 2000,
            children: {
              'test-incident-2_my-class-1_alpha': {
                id: 'test-incident-2_my-class-1_alpha',
                position: 0,
                duration: 2000
              },
              'test-incident-2_my-class-1_beta': {
                id: 'test-incident-2_my-class-1_beta',
                position: 0,
                duration: 2000
              }
            }
          }
        }
      },
      'test-group-1': {
        id: 'test-group-1',
        position: 1000,
        duration: 3000,
        children: {
          'test-incident-1': {
            // ElementSplitter
            id: 'test-incident-1',
            position: 1000,
            duration: 2000,
            children: {
              'test-incident-1_my-class-1': {
                // AnimatedAttributeSplitter
                id: 'test-incident-1_my-class-1',
                position: 0,
                duration: 2000,
                children: {
                  'test-incident-1_my-class-1_alpha': {
                    // ContextAwareIncidents
                    id: 'test-incident-1_my-class-1_alpha',
                    position: 0,
                    duration: 2000
                  },
                  'test-incident-1_my-class-1_beta': {
                    id: 'test-incident-1_my-class-1_beta',
                    position: 0,
                    duration: 2000
                  }
                }
              },
              'test-incident-1_my-class-2': {
                id: 'test-incident-1_my-class-2',
                position: 0,
                duration: 2000,
                children: {
                  'test-incident-1_my-class-2_alpha': {
                    id: 'test-incident-1_my-class-2_alpha',
                    position: 0,
                    duration: 2000
                  },
                  'test-incident-1_my-class-2_beta': {
                    id: 'test-incident-1_my-class-2_beta',
                    position: 0,
                    duration: 2000
                  }
                }
              }
            }
          }
        }
      }
    }
  }
};
