// @kissmybutton/motorcortex/dist/DescriptiveIncidents/__tests__/states/resize_state_8.js
"use strict";

module.exports = {
  descriptiveTree: {
    id: 'my-clip',
    duration: 1400,
    children: {
      'desc-group-1': {
        id: 'desc-group-1',
        duration: 1200,
        position: 200,
        children: {
          'desc-inc-1': {
            id: 'desc-inc-1',
            position: 25,
            duration: 50
          },
          'desc-group-2': {
            id: 'desc-group-2',
            position: 150,
            duration: 540,
            children: {
              'desc-inc-2': {
                id: 'desc-inc-2',
                position: 40,
                duration: 500
              }
            }
          },
          'casi': {
            id: 'casi',
            duration: 1200,
            position: 0
          }
        }
      }
    }
  },
  lanes: {
    'my-class-1___alpha': [{
      id: 'desc-inc-1_my-class-1_alpha',
      millisecond: 225,
      initialValue: 0
    }, {
      id: 'desc-inc-2_my-class-1_alpha',
      millisecond: 390,
      initialValue: 2
    }],
    'my-class-2___alpha': [{
      id: 'desc-inc-1_my-class-2_alpha',
      millisecond: 225,
      initialValue: 0
    }, {
      id: 'desc-inc-2_my-class-2_alpha',
      millisecond: 390,
      initialValue: 2
    }]
  },
  realTree: {
    id: 'my-clip',
    duration: 1400,
    children: {
      'desc-group-1': {
        id: 'desc-group-1',
        position: 200,
        duration: 1200,
        children: {
          'desc-inc-1': {
            id: 'desc-inc-1',
            position: 25,
            duration: 50,
            children: {
              'desc-inc-1_my-class-1': {
                id: 'desc-inc-1_my-class-1',
                position: 0,
                duration: 50,
                children: {
                  'desc-inc-1_my-class-1_alpha': {
                    id: 'desc-inc-1_my-class-1_alpha',
                    position: 0,
                    duration: 50
                  }
                }
              },
              'desc-inc-1_my-class-2': {
                id: 'desc-inc-1_my-class-2',
                position: 0,
                duration: 50,
                children: {
                  'desc-inc-1_my-class-2_alpha': {
                    id: 'desc-inc-1_my-class-2_alpha',
                    position: 0,
                    duration: 50
                  }
                }
              }
            }
          },
          'desc-group-2': {
            id: 'desc-group-2',
            position: 150,
            duration: 540,
            children: {
              'desc-inc-2': {
                id: 'desc-inc-2',
                position: 40,
                duration: 500,
                children: {
                  'desc-inc-2_my-class-1': {
                    id: 'desc-inc-2_my-class-1',
                    position: 0,
                    duration: 500,
                    children: {
                      'desc-inc-2_my-class-1_alpha': {
                        id: 'desc-inc-2_my-class-1_alpha',
                        position: 0,
                        duration: 500
                      }
                    }
                  },
                  'desc-inc-2_my-class-2': {
                    id: 'desc-inc-2_my-class-2',
                    position: 0,
                    duration: 500,
                    children: {
                      'desc-inc-2_my-class-2_alpha': {
                        id: 'desc-inc-2_my-class-2_alpha',
                        position: 0,
                        duration: 500
                      }
                    }
                  }
                }
              }
            }
          },
          'casi': {
            id: 'casi',
            position: 0,
            duration: 1200
          }
        }
      }
    }
  }
};
