// @kissmybutton/motorcortex/dist/Channels/__tests__/states/iv_state8.js
"use strict";

module.exports = {
  lanes: {
    'my-class-1___alpha': [{
      id: 'incident2_my-class-1_alpha',
      millisecond: 1000,
      initialValue: 1,
      targetValue: 20
    }, {
      id: 'incident1_my-class-1_alpha',
      millisecond: 3000,
      initialValue: 3,
      targetValue: 2
    }],
    'my-class-1___beta': [{
      id: 'incident2_my-class-1_beta',
      millisecond: 1000,
      initialValue: 2.5,
      targetValue: 20
    }, {
      id: 'incident1_my-class-1_beta',
      millisecond: 3000,
      initialValue: 20,
      targetValue: 2
    }],
    'my-class-1___compo': [{
      id: 'cincident2_my-class-1_compo',
      millisecond: 1000,
      initialValue: {
        gamma: 3,
        delta: 9,
        epsilon: 10
      },
      targetValue: {
        gamma: 10,
        delta: 11,
        epsilon: 12
      }
    }, {
      id: 'cincident1_my-class-1_compo',
      millisecond: 3000,
      initialValue: {
        gamma: 0,
        delta: 11,
        epsilon: 1
      },
      targetValue: {
        gamma: 7,
        delta: 8,
        epsilon: 1
      }
    }]
  }
};
