// @kissmybutton/motorcortex/dist/Channels/__tests__/states/iv_state10.js
"use strict";

module.exports = {
  lanes: {
    'my-class-1___alpha': [{
      id: 'incident1_my-class-1_alpha',
      millisecond: 3000,
      initialValue: 3
    }, {
      id: 'incident2_my-class-1_alpha',
      millisecond: 4000,
      initialValue: 2
    }, {
      id: 'incident3_my-class-1_alpha',
      millisecond: 5000,
      initialValue: 20
    }],
    'my-class-1___beta': [{
      id: 'incident1_my-class-1_beta',
      millisecond: 3000,
      initialValue: 2.5
    }, {
      id: 'incident2_my-class-1_beta',
      millisecond: 4000,
      initialValue: 2
    }, {
      id: 'incident3_my-class-1_beta',
      millisecond: 5000,
      initialValue: 15
    }],
    'my-class-1___compo': [{
      id: 'cincident1_my-class-1_compo',
      millisecond: 3000,
      initialValue: {
        gamma: 0,
        delta: 4,
        epsilon: 1
      },
      targetValue: {
        gamma: 7,
        delta: 8,
        epsilon: 1
      }
    }, {
      id: 'cincident2_my-class-1_compo',
      millisecond: 4000,
      initialValue: {
        gamma: 7,
        delta: 9,
        epsilon: 10
      },
      targetValue: {
        gamma: 10,
        delta: 11,
        epsilon: 12
      }
    }, {
      id: 'cincident3_my-class-1_compo',
      millisecond: 5000,
      initialValue: {
        gamma: 10,
        delta: 11,
        epsilon: 12
      },
      targetValue: {
        gamma: 30,
        delta: 11,
        epsilon: 12
      }
    }]
  }
};
