// @kissmybutton/motorcortex/dist/Channels/__tests__/states/state2.js
"use strict";

module.exports = {
  descriptiveTree: {
    id: 'my-clip',
    duration: 6000,
    children: {
      'di2': {
        id: 'di2',
        position: 0,
        duration: 1000
      },
      'di1': {
        id: 'di1',
        position: 1000,
        duration: 1000
      },
      'di3': {
        id: 'di3',
        position: 5000,
        duration: 1000
      }
    }
  },
  lanes: {
    'my-class-1___alpha': [{
      id: 'di2_my-class-1_alpha',
      millisecond: 0,
      initialValue: 0
    }, {
      id: 'di1_my-class-1_alpha',
      millisecond: 1000,
      initialValue: 2
    }, {
      id: 'di3_my-class-1_alpha',
      millisecond: 5000,
      initialValue: 1
    }]
  },
  realTree: {
    id: 'my-clip',
    duration: 6000,
    children: {
      'di2': {
        id: 'di2',
        position: 0,
        duration: 1000,
        children: {
          'di2_my-class-1': {
            id: 'di2_my-class-1',
            position: 0,
            duration: 1000,
            children: {
              'di2_my-class-1_alpha': {
                id: 'di2_my-class-1_alpha',
                position: 0,
                duration: 1000
              }
            }
          }
        }
      },
      'di1': {
        id: 'di1',
        position: 1000,
        duration: 1000,
        children: {
          'di1_my-class-1': {
            id: 'di1_my-class-1',
            position: 0,
            duration: 1000,
            children: {
              'di1_my-class-1_alpha': {
                id: 'di1_my-class-1_alpha',
                position: 0,
                duration: 1000
              }
            }
          }
        }
      },
      'di3': {
        id: 'di3',
        position: 5000,
        duration: 1000,
        children: {
          'di3_my-class-1': {
            id: 'di3_my-class-1',
            position: 0,
            duration: 1000,
            children: {
              'di3_my-class-1_alpha': {
                id: 'di3_my-class-1_alpha',
                position: 0,
                duration: 1000
              }
            }
          }
        }
      }
    }
  }
};
