// @kissmybutton/motorcortex/src/DescriptiveIncidents/__tests__/states/dc12.js
module.exports = {
    descriptiveTree: {
        id: 'my-clip',
        duration: 11000,
        children: {
            'my-group': {
                id: 'my-group',
                position: 200,
                duration: 0,
                children: {}
            },
            'test-incident-2': {
                id: 'test-incident-2',
                position: 5000,
                duration: 6000
            },
            'test-group-1': {
                id: 'test-group-1',
                position: 1000,
                duration: 9000,
                children: {
                    'test-incident-1': {
                        id: 'test-incident-1',
                        position: 7000,
                        duration: 2000
                    }
                }
            }

        }
    },
    lanes: {
        'my-class-1___alpha': [{
            id: 'test-incident-1_my-class-1_alpha',
            millisecond: 8000,
            initialValue: 1
        }],
        'my-class-1___beta': [{
            id: 'test-incident-1_my-class-1_beta',
            millisecond: 8000,
            initialValue: 0
        }],
        'my-class-1___gamma': [{
            id: 'test-incident-2_my-class-1_gamma',
            millisecond: 5000,
            initialValue: 0
        }],
        'my-class-2___alpha': [{
            id: 'test-incident-1_my-class-2_alpha',
            millisecond: 8000,
            initialValue: 1
        }],
        'my-class-2___beta': [{
            id: 'test-incident-1_my-class-2_beta',
            millisecond: 8000,
            initialValue: 0
        }]
    },
    realTree: {
        id: 'my-clip',
        duration: 11000,
        children: {
            'my-group': {
                id: 'my-group',
                position: 200,
                duration: 0,
                children: {}
            },
            'test-incident-2': {
                id: 'test-incident-2',
                position: 5000,
                duration: 6000,
                children: {
                    'test-incident-2_my-class-1': {
                        id: 'test-incident-2_my-class-1',
                        position: 0,
                        duration: 6000,
                        children: {
                            'test-incident-2_my-class-1_gamma': {
                                id: 'test-incident-2_my-class-1_gamma',
                                position: 0,
                                duration: 6000
                            }
                        }
                    }
                }
            },
            'test-group-1': {
                id: 'test-group-1',
                position: 1000,
                duration: 9000,
                children: {
                    'test-incident-1': { // ElementSplitter
                        id: 'test-incident-1',
                        position: 7000,
                        duration: 2000,
                        children: {
                            'test-incident-1_my-class-1': { // AnimatedAttributeSplitter
                                id: 'test-incident-1_my-class-1',
                                position: 0,
                                duration: 2000,
                                children: {
                                    'test-incident-1_my-class-1_alpha': { // ContextAwareIncidents
                                        id: 'test-incident-1_my-class-1_alpha',
                                        position: 0,
                                        duration: 2000
                                    },
                                    'test-incident-1_my-class-1_beta': {
                                        id: 'test-incident-1_my-class-1_beta',
                                        position: 0,
                                        duration: 2000
                                    }
                                }
                            },
                            'test-incident-1_my-class-2': {
                                id: 'test-incident-1_my-class-2',
                                position: 0,
                                duration: 2000,
                                children: {
                                    'test-incident-1_my-class-2_alpha': {
                                        id: 'test-incident-1_my-class-2_alpha',
                                        position: 0,
                                        duration: 2000
                                    },
                                    'test-incident-1_my-class-2_beta': {
                                        id: 'test-incident-1_my-class-2_beta',
                                        position: 0,
                                        duration: 2000
                                    }
                                }
                            }
                        }
                    }
                }
            }

        }
    }
}
