// @kissmybutton/motorcortex/src/DescriptiveIncidents/__tests__/states/dcasi1.js
module.exports = {
    descriptiveTree: {
        id: 'my-root-clip',
        duration: 4500,
        children: {
            'casi': {
                id: 'casi',
                position: 500,
                duration: 4000
            }
        }
    },
    clipsLane: [{
        millisecond: 500,
        id: 'casi_my-class-1'
    }, {
        millisecond: 500,
        id: 'casi_my-class-2'
    }],
    realTree: {
        id: 'my-root-clip',
        duration: 4500,
        children: {
            'casi': {
                id: 'casi', // dom aware clip incident
                position: 500,
                duration: 4000,
                children: {
                    'casi_my-class-1': {
                        id: 'casi_my-class-1',
                        position: 0,
                        duration: 4000
                    },
                    'casi_my-class-2': {
                        id: 'casi_my-class-2',
                        position: 0,
                        duration: 4000
                    }
                }
            }

        }
    }
}
