// @kissmybutton/motorcortex/src/DescriptiveIncidents/__tests__/states/resize_state_9.js
module.exports = {
    descriptiveTree: {
        id: 'my-clip',
        duration: 2600,
        children: {
            'desc-group-1': {
                id: 'desc-group-1',
                duration: 2400,
                position: 200,
                children: {
                    'desc-inc-1': {
                        id: 'desc-inc-1',
                        position: 50,
                        duration: 100
                    },
                    'desc-group-2': {
                        id: 'desc-group-2',
                        position: 300,
                        duration: 1080,
                        children: {
                            'desc-inc-2': {
                                id: 'desc-inc-2',
                                position: 80,
                                duration: 1000
                            }
                        }
                    },
                    'casi': {
                        id: 'casi',
                        duration: 2400,
                        position: 0
                    }
                }
            }
        }
    },
    lanes: {
        'my-class-1___alpha': [{
                id: 'desc-inc-1_my-class-1_alpha',
                millisecond: 250,
                initialValue: 0
            },
            {
                id: 'desc-inc-2_my-class-1_alpha',
                millisecond: 580,
                initialValue: 2
            }
        ],
        'my-class-2___alpha': [{
                id: 'desc-inc-1_my-class-2_alpha',
                millisecond: 250,
                initialValue: 0
            },
            {
                id: 'desc-inc-2_my-class-2_alpha',
                millisecond: 580,
                initialValue: 2
            }
        ]
    },
    realTree: {
        id: 'my-clip',
        duration: 2600,
        children: {
            'desc-group-1': {
                id: 'desc-group-1',
                position: 200,
                duration: 2400,
                children: {
                    'desc-inc-1': {
                        id: 'desc-inc-1',
                        position: 50,
                        duration: 100,
                        children: {
                            'desc-inc-1_my-class-1': {
                                id: 'desc-inc-1_my-class-1',
                                position: 0,
                                duration: 100,
                                children: {
                                    'desc-inc-1_my-class-1_alpha': {
                                        id: 'desc-inc-1_my-class-1_alpha',
                                        position: 0,
                                        duration: 100
                                    }
                                }
                            },
                            'desc-inc-1_my-class-2': {
                                id: 'desc-inc-1_my-class-2',
                                position: 0,
                                duration: 100,
                                children: {
                                    'desc-inc-1_my-class-2_alpha': {
                                        id: 'desc-inc-1_my-class-2_alpha',
                                        position: 0,
                                        duration: 100
                                    }
                                }
                            }
                        }
                    },
                    'desc-group-2': {
                        id: 'desc-group-2',
                        position: 300,
                        duration: 1080,
                        children: {
                            'desc-inc-2': {
                                id: 'desc-inc-2',
                                position: 80,
                                duration: 1000,
                                children: {
                                    'desc-inc-2_my-class-1': {
                                        id: 'desc-inc-2_my-class-1',
                                        position: 0,
                                        duration: 1000,
                                        children: {
                                            'desc-inc-2_my-class-1_alpha': {
                                                id: 'desc-inc-2_my-class-1_alpha',
                                                position: 0,
                                                duration: 1000
                                            }
                                        }
                                    },
                                    'desc-inc-2_my-class-2': {
                                        id: 'desc-inc-2_my-class-2',
                                        position: 0,
                                        duration: 1000,
                                        children: {
                                            'desc-inc-2_my-class-2_alpha': {
                                                id: 'desc-inc-2_my-class-2_alpha',
                                                position: 0,
                                                duration: 1000
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    },
                    'casi': {
                        id: 'casi',
                        position: 0,
                        duration: 2400
                    }
                }
            }
        }
    }
}
