// @kissmybutton/motorcortex/src/DescriptiveIncidents/__tests__/states/resize_state_9-sub.js
module.exports = {
    descriptiveTree: {
        id: 'casi',
        duration: 2400,
        children: {
            'casi-incident': {
                id: 'casi-incident',
                duration: 1600,
                position: 800
            }
        }
    },
    lanes: {
        'my-class-1___alpha': [{
            id: 'casi-incident_my-class-1_alpha',
            millisecond: 800,
            initialValue: 0
        }],
        'my-class-2___alpha': [{
            id: 'casi-incident_my-class-2_alpha',
            millisecond: 800,
            initialValue: 0
        }]
    },
    realTree: {
        id: 'casi',
        duration: 2400,
        children: {
            'casi-incident': {
                id: 'casi-incident',
                position: 800,
                duration: 1600,
                children: {
                    'casi-incident_my-class-1': {
                        id: 'casi-incident_my-class-1',
                        position: 0,
                        duration: 1600,
                        children: {
                            'casi-incident_my-class-1_alpha': {
                                id: 'casi-incident_my-class-1_alpha',
                                position: 0,
                                duration: 1600
                            }
                        }
                    },
                    'casi-incident_my-class-2': {
                        id: 'casi-incident_my-class-2',
                        position: 0,
                        duration: 1600,
                        children: {
                            'casi-incident_my-class-2_alpha': {
                                id: 'casi-incident_my-class-2_alpha',
                                position: 0,
                                duration: 1600
                            }
                        }
                    }
                }
            }
        }
    }
}
