// @kissmybutton/motorcortex/src/DescriptiveIncidents/__tests__/states/resize_state_5.js
module.exports = {
    descriptiveTree: {
        id: 'my-clip',
        duration: 4450,
        children: {
            'desc-group-1': {
                id: 'desc-group-1',
                duration: 3450,
                position: 1000,
                children: {
                    'desc-inc-1': {
                        id: 'desc-inc-1',
                        position: 125,
                        duration: 250
                    },
                    'desc-group-2': {
                        id: 'desc-group-2',
                        position: 750,
                        duration: 2700,
                        children: {
                            'desc-inc-2': {
                                id: 'desc-inc-2',
                                position: 200,
                                duration: 2500
                            }
                        }
                    }
                }
            },
            'casi': {
                id: 'casi',
                duration: 3000,
                position: 1000,
            }
        }
    },
    lanes: {
        'my-class-1___alpha': [{
                id: 'desc-inc-1_my-class-1_alpha',
                millisecond: 1125,
                initialValue: 0
            },
            {
                id: 'desc-inc-2_my-class-1_alpha',
                millisecond: 1950,
                initialValue: 2
            }
        ],
        'my-class-2___alpha': [{
                id: 'desc-inc-1_my-class-2_alpha',
                millisecond: 1125,
                initialValue: 0
            },
            {
                id: 'desc-inc-2_my-class-2_alpha',
                millisecond: 1950,
                initialValue: 2
            }
        ]
    },
    realTree: {
        id: 'my-clip',
        duration: 4450,
        children: {
            'desc-group-1': {
                id: 'desc-group-1',
                position: 1000,
                duration: 3450,
                children: {
                    'desc-inc-1': {
                        id: 'desc-inc-1',
                        position: 125,
                        duration: 250,
                        children: {
                            'desc-inc-1_my-class-1': {
                                id: 'desc-inc-1_my-class-1',
                                position: 0,
                                duration: 250,
                                children: {
                                    'desc-inc-1_my-class-1_alpha': {
                                        id: 'desc-inc-1_my-class-1_alpha',
                                        position: 0,
                                        duration: 250
                                    }
                                }
                            },
                            'desc-inc-1_my-class-2': {
                                id: 'desc-inc-1_my-class-2',
                                position: 0,
                                duration: 250,
                                children: {
                                    'desc-inc-1_my-class-2_alpha': {
                                        id: 'desc-inc-1_my-class-2_alpha',
                                        position: 0,
                                        duration: 250
                                    }
                                }
                            }
                        }
                    },
                    'desc-group-2': {
                        id: 'desc-group-2',
                        position: 750,
                        duration: 2700,
                        children: {
                            'desc-inc-2': {
                                id: 'desc-inc-2',
                                position: 200,
                                duration: 2500,
                                children: {
                                    'desc-inc-2_my-class-1': {
                                        id: 'desc-inc-2_my-class-1',
                                        position: 0,
                                        duration: 2500,
                                        children: {
                                            'desc-inc-2_my-class-1_alpha': {
                                                id: 'desc-inc-2_my-class-1_alpha',
                                                position: 0,
                                                duration: 2500
                                            }
                                        }
                                    },
                                    'desc-inc-2_my-class-2': {
                                        id: 'desc-inc-2_my-class-2',
                                        position: 0,
                                        duration: 2500,
                                        children: {
                                            'desc-inc-2_my-class-2_alpha': {
                                                id: 'desc-inc-2_my-class-2_alpha',
                                                position: 0,
                                                duration: 2500
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            },
            'casi': {
                id: 'casi',
                position: 1000,
                duration: 3000
            }
        }
    }
}
