// @kissmybutton/motorcortex/src/DescriptiveIncidents/__tests__/states/resize_state_7-sub.js
export const descriptiveTree = {
  id: 'casi',
  duration: 1200,
  children: {
    'casi-incident': {
      id: 'casi-incident',
      duration: 800,
      position: 400
    }
  }
};
export const lanes = {
  'my-class-1___alpha': [
    {
      id: 'casi-incident_my-class-1_alpha',
      millisecond: 400,
      initialValue: 0
    }
  ],
  'my-class-2___alpha': [
    {
      id: 'casi-incident_my-class-2_alpha',
      millisecond: 400,
      initialValue: 0
    }
  ]
};
export const realTree = {
  id: 'casi',
  duration: 1200,
  children: {
    'casi-incident': {
      id: 'casi-incident',
      position: 400,
      duration: 800,
      children: {
        'casi-incident_my-class-1': {
          id: 'casi-incident_my-class-1',
          position: 0,
          duration: 800,
          children: {
            'casi-incident_my-class-1_alpha': {
              id: 'casi-incident_my-class-1_alpha',
              position: 0,
              duration: 800
            }
          }
        },
        'casi-incident_my-class-2': {
          id: 'casi-incident_my-class-2',
          position: 0,
          duration: 800,
          children: {
            'casi-incident_my-class-2_alpha': {
              id: 'casi-incident_my-class-2_alpha',
              position: 0,
              duration: 800
            }
          }
        }
      }
    }
  }
};
