// @kissmybutton/motorcortex/src/Audio/helpers/AudioNode.js
import { audioContext } from '../../_coreUtils/Helper';

class AudioNode {
  constructor() {
    // audio context initialisation

    this.output = audioContext.createGain();
    this.gainNode = audioContext.createGain();
    if (audioContext.createStereoPanner) {
      this.pannerNode = audioContext.createStereoPanner(); // if it is supported by the browser
    }

    // this.highpassNode = audioContext.createBiquadFilter();
    // this.lowpassNode = audioContext.createBiquadFilter();

    // this.lowpassNode.connect(this.highpassNode);
    // this.highpassNode.connect(this.pannerNode);
    if (audioContext.createStereoPanner) {
      this.pannerNode.connect(this.gainNode);
      this.gainNode.connect(this.output);
      this.input = this.pannerNode;
    } else {
      this.gainNode.connect(this.output);
      this.input = this.gainNode;
    }
  }

  connect(master) {
    this.output.connect(master);
  }

  disconnect() {
    // eslint-disable-line no-unused-vars
    this.output.disconnect();
  }
}

export default AudioNode;
