// @kissmybutton/motorcortex/src/Audio/helpers/AudioNode.js
const Helper = require('../../_coreUtils/Helper');
const helper = new Helper();

class AudioNode {
    constructor() {
        // audio context initialisation
        const audioContext = helper.getAudioContext();

        this.output = audioContext.createGain();
        this.gainNode = audioContext.createGain();
        if (audioContext.createStereoPanner) {
            this.pannerNode = audioContext.createStereoPanner(); // if it is supported by the browser
        }

        // this.highpassNode = audioContext.createBiquadFilter();
        // this.lowpassNode = audioContext.createBiquadFilter();

        // this.lowpassNode.connect(this.highpassNode);
        // this.highpassNode.connect(this.pannerNode);
        if (audioContext.createStereoPanner) {
            this.pannerNode.connect(this.gainNode);
            this.gainNode.connect(this.output);
            this.input = this.pannerNode;
        }
        else {
            this.gainNode.connect(this.output);
            this.input = this.gainNode;
        }
    }

    connect(master) {
        this.output.connect(master);
    }

    disconnect(master) {
        this.output.disconnect();
    }
}

module.exports = AudioNode;
