// @kissmybutton/motorcortex/src/Channels/__tests__/states/iv_state9.js
module.exports = {
    lanes: {
        'my-class-1___alpha': [{
            id: 'incident1_my-class-1_alpha',
            millisecond: 3000,
            initialValue: 3,
            targetValue: 2
        }, {
            id: 'incident2_my-class-1_alpha',
            millisecond: 4000,
            initialValue: 2,
            targetValue: 20
        }, {
            id: 'incident3_my-class-1_alpha',
            millisecond: 5000,
            initialValue: 20,
            targetValue: 30
        }],
        'my-class-1___beta': [{
            id: 'incident1_my-class-1_beta',
            millisecond: 3000,
            initialValue: 2.5,
            targetValue: 2
        }, {
            id: 'incident2_my-class-1_beta',
            millisecond: 4000,
            initialValue: 2,
            targetValue: 20
        }, {
            id: 'incident3_my-class-1_beta',
            millisecond: 5000,
            initialValue: 15,
            targetValue: 30
        }],
        'my-class-1___compo': [{
            id: 'cincident2_my-class-1_compo',
            millisecond: 1000,
            initialValue: {
                gamma: 3,
                delta: 9,
                epsilon: 10
            },
            targetValue: {
                gamma: 10,
                delta: 11,
                epsilon: 12
            }
        }, {
            id: 'cincident1_my-class-1_compo',
            millisecond: 3000,
            initialValue: {
                gamma: 0,
                delta: 11,
                epsilon: 1
            },
            targetValue: {
                gamma: 7,
                delta: 8,
                epsilon: 1
            }
        }]
    }
}
