// @kissmybutton/motorcortex/src/Channels/__tests__/states/state5.js
module.exports = {
    descriptiveTree: {
        id: 'my-clip',
        duration: 6000,
        children: {
            'cdi3': {
                id: 'cdi3',
                position: 0,
                duration: 1000
            },
            'cdi2': {
                id: 'cdi2',
                position: 3000,
                duration: 1000
            },
            'cdi1': {
                id: 'cdi1',
                position: 5000,
                duration: 1000
            }

        }
    },
    lanes: {
        'my-class-1___compo': [{
            id: 'cdi3_my-class-1_compo',
            millisecond: 0,
            initialValue: {
                gamma: 0,
                delta: 0,
                epsilon: 0
            }
        }, {
            id: 'cdi2_my-class-1_compo',
            millisecond: 3000,
            initialValue: {
                gamma: 3,
                delta: 3,
                epsilon: 0
            }
        }, {
            id: 'cdi1_my-class-1_compo',
            millisecond: 5000,
            initialValue: {
                gamma: 3,
                delta: 2,
                epsilon: 0
            }
        }]
    },
    realTree: {
        id: 'my-clip',
        duration: 6000,
        children: {
            'cdi3': {
                id: 'cdi3',
                position: 0,
                duration: 1000,
                children: {
                    'cdi3_my-class-1': {
                        id: 'cdi3_my-class-1',
                        position: 0,
                        duration: 1000,
                        children: {
                            'cdi3_my-class-1_compo': {
                                id: 'cdi3_my-class-1_compo',
                                position: 0,
                                duration: 1000
                            }
                        }
                    }
                }
            },
            'cdi2': {
                id: 'cdi2',
                position: 3000,
                duration: 1000,
                children: {
                    'cdi2_my-class-1': {
                        id: 'cdi2_my-class-1',
                        position: 0,
                        duration: 1000,
                        children: {
                            'cdi2_my-class-1_compo': {
                                id: 'cdi2_my-class-1_compo',
                                position: 0,
                                duration: 1000
                            }
                        }
                    }
                }
            },
            'cdi1': {
                id: 'cdi1',
                position: 5000,
                duration: 1000,
                children: {
                    'cdi1_my-class-1': {
                        id: 'cdi1_my-class-1',
                        position: 0,
                        duration: 1000,
                        children: {
                            'cdi1_my-class-1_compo': {
                                id: 'cdi1_my-class-1_compo',
                                position: 0,
                                duration: 1000
                            }
                        }
                    }
                }
            }

        }
    }
}
