// @kissmybutton/motorcortex/pluginMainTemplate.js
const YourIncident = require('./path/to/your/class-definition');
const YourSelfContainedIncident = require('./path/to/your/class-definition');
const YourChannel = require('./path/to/your/channel')

module.exports = {
    npm_name: "the-npm-name-of-your-plugin",
    incidents: [
        {
            exportable: YourIncident,
            attributes: {
                animated: [
                    {
                        key: oneof ['left', 'top']
                    }
                ],
                plain: [
                    {
                        key: 'frequency',
                        type: 'integer',
                        acceptsMinor: false,
                        acceptsZero: false,
                        madatory: false,
                        default: 30,
                        minValue: 0,
                        maxValue: 2000,
                        uiItem: 'input'
                    }
                ]
            }
        },
        {
            exportable: YourSelfContainedIncident,
            containerParams: {
                width: "500px",
                height: "400px",
                cssIsolation: "isolate" // either isolate (default) or open
            }
        }
    ],
    channel: YourChannel
};
